docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+    .option("port", 9999)
+    .load()
+
+socketDF.isStreaming    // Returns true for DataFrames that has a streaming source
+
+socketDF.printSchema 
+
+// Read all the csv files written atomically in a directory
+val userSchema = new StructType().add("name", "string").add("age", "integer")
+val csvDF = spark
+    .readStream
+    .option("sep", ";")
+    .schema(userSchema)      // Specify schema of the parquet files
+    .csv("/path/to/directory")    // Equivalent to format("cv").load("/path/to/directory")
+{% endhighlight %}
+
+</div>
+<div data-lang="java"  markdown="1">
+
+{% highlight java %}
+SparkSession spark = ...
+
+// Read text from socket 
+Dataset[Row] socketDF = spark
+    .readStream()
+    .format("socket")
+    .option("host", "localhost")
+    .option("port", 9999)
+    .load();
+
+socketDF.isStreaming();    // Returns True for DataFrames that has a streaming source
+
+socketDF.printSchema();
+
+// Read all the csv files written atomically in a directory
+StructType userSchema = new StructType().add("name", "string").add("age", "integer");
+Dataset[Row] csvDF = spark
+    .readStream()
+    .option("sep", ";")
+    .schema(userSchema)      // Specify schema of the parquet files
+    .csv("/path/to/directory");    // Equivalent to format("cv").load("/path/to/directory")
+{% endhighlight %}
+
+</div>
+<div data-lang="python"  markdown="1">
+
+{% highlight python %}
+spark = SparkSession. …. 
+
+# Read text from socket 
+socketDF = spark \
+    .readStream()  \
+    .format("socket") \
+    .option("host", "localhost") \
+    .option("port", 9999) \
+    .load()
+
+socketDF.isStreaming()    # Returns True for DataFrames that has a streaming source
+
+socketDF.printSchema() 
+
+# Read all the csv files written atomically in a directory
+userSchema = StructType().add("name", "string").add("age", "integer")
+csvDF = spark \
+    .readStream() \
+    .option("sep", ";") \
+    .schema(userSchema) \
+    .csv("/path/to/directory")    # Equivalent to format("cv").load("/path/to/directory")
+{% endhighlight %}
+
+</div>
+</div>
+
+These examples generate streaming DataFrames that are untyped, that is, the schema of the DataFrame is not checked at compile time, only checked in runtime when the query is submitted. Some operations like `map`, `flatMap`, etc. need the type to be known at compile time. To do those, you can convert these untyped streaming DataFrames to typed streaming Datasets using the same methods as static DataFrame. See the SQL Programming Guide for more details. Additionally, more details on the supported streaming sources are discussed later in the document.
+
+## Operations on streaming DataFrames/Datasets
+You can apply all kinds of streaming DataFrames/Datasets - ranging for untyped, SQL-like operations (e.g. select, where, groupBy), to typed RDD-like operations (e.g. map, filter, flatMap). See the SQL programming guide for more details. Let’s take a look at a few example operations that you can use.
```

Comment:
all kinds of streaming DataFrames/Datasets - ranging for =>
all kinds of operations to streaming DataFrames/Datasets - ranging from